Anonymous FTP request single file prompt (vs 2)

kbrock opened the pull request:
We were prompting for a local data filename and a remote url.
But since this is streaming, it does not make sense.

This now:

- properly populates the `:remote_file_name value`
- hides/shows the prompt selectively for dump, backup, restore
- has configurable rake options

related to:
- https://bugzilla.redhat.com/show_bug.cgi?id=1632433
- ManageIQ/manageiq-appliance_console#65

configuration has been tweaked:

```yaml
  database_admin:
    menu_order:
    - local
    - ftp://ftp.example.com/incoming/
    local: Local file
    prompts:
      ftp.example.com:
        enabled_for: ["backup", "dump"]
        filename_text: "The case number dash (-) filename. (e.g.: 12345-db.backup)"
        filename_validator: "^[0-9]{4,}-..*"
        rake_options:
          skip_directory: true
```

NickLaMuro commented on lib/manageiq/appliance_console/database_admin.rb:
```diff
@@ -140,14 +140,12 @@ def ask_ftp_file_options
-        uri_filename = ask_custom_prompt(hostname, 'filename', "Target filename (e.g.: #{sample_case})")
-        @uri         = server_uri.gsub(sample_case, uri_filename)
+        hostname  = URI(server_uri).host
+        @filename = ask_custom_prompt(hostname, 'filename', "Target filename for backup")
```
I realize this was added in the previous PR, and I wasn't really the reviewer, but why are we "passing" the default prompt if this method is only used in one spot?

I was reviewing a similar PR that @carbonin was (where he asked to axe some constants), but this seems like it makes more sense to just have this as a constant, or a frozen string in the method itself instead of passing it from here.

NickLaMuro replied:
Also (not part of this line of code), but I think you can delete the constants at the top of this file for `LOCAL_FILE`, `NFS_FILE`, etc. since we don't use them any more (as of #65).

Separate commit/PR maybe?  I think you would have a lot of tests that would need to be updated, for what it is worth.

kbrock replied:
ok, will make the prompt less generic.

Those constants are used all over the specs :(  They have been simplified quite a bit (now just have `'ftp'` vs the full text string) -- we can hardcode phrases like "ftp" in the specs.

NickLaMuro commented on spec/database_admin_spec.rb:
```diff
@@ -1736,38 +1736,32 @@ def confirm_and_execute
     end
 
     describe "#ask_custom_file_options" do
```
Looking back at `describe "#ask_custom_file_options" do` in the last PR, you only added this for the current context this is under (I think just "DB Dump").

Can you add specs for the other two?  (you should be able to grep for the above to find the other places where this is at).

kbrock commented:
ok

- added the missing tests (they felt a little copy paste),
- fixed custom_file prompt (removed the genericness)
- removed the constants (that were only used in specs)

keep 'em comin'

kbrock commented:
@NickLaMuro What I used to configure the appliance console in local development to act like the appliance vm and pickup my localization changes

```
$ pwd
/Users/kbrock/src/manageiq-appliance_console
$ cat ~/src/manageiq/productization/appliance_console/locales/appliance/en_productization.yml
---
en:
  database_admin:
    menu_order:
    - local
    - ftp://ftp.example.com/incoming
    local: Local file
    prompts:
      ftp.example.com:
        filename_text: "The case number dash (-) filename. (e.g.: 12345-db.backup)"
        filename_validator: "^[0-9]{4,}-..*"
$ bundle exec irb -I lib
irb> module ManageIQ; module ApplianceConsole; RAILS_ROOT = File.expand_path("../manageiq", Dir.pwd); end; end
# => "/Users/kbrock/src/manageiq"  
irb> require 'manageiq/appliance_console/i18n'
# => true  
irb> I18n.t("database_admin.prompts", :default => {})
# => {
  :"ftp.example.com"=>{
    :filename_text=>"The case number dash (-) filename. (e.g.: 12345-db.backup)",
    :filename_validator=>"^[0-9]{4,}-..*"
  }
}  
irb> I18n.load_path
# => [
  "/Users/kbrock/src/manageiq-appliance_console/locales/appliance/en.yml",
  "/Users/kbrock/src/manageiq/productization/appliance_console/locales/appliance/en_productization.yml"
]  
```

NickLaMuro commented on lib/manageiq/appliance_console/database_admin.rb:
```diff
@@ -140,21 +138,22 @@ def ask_ftp_file_options
-        params[:remote_file_name] = filename if filename
+        params = {
+          :uri              => uri,
+          :skip_directory   => true,
```
I am fine with this for now, but I think in the end, this should be part of the configuration, and not hard coded in.

NickLaMuro replied:
Along with this, I really feel we should be hiding this option, or allowing it to be configured to do so, for the restore options since this is ultimately being implemented for an endpoint you won't be able to restore from.

Might need a follow up, but just my two cents.

kbrock commented:
WIPping to find a more configurable way of adding flags for specific backup options

kbrock commented:
database_admin:
    menu_order:
    - local
    - ftp://ftp.example.com/incoming/
    local: Local file
    prompts:
      ftp.example.com:
        filename_text: "The case number dash (-) filename. (e.g.: 12345-db.backup)"
        filename_validator: "^[0-9]{4,}-..*"

kbrock commented:
> Along with this, I really feel we should be hiding this option, or allowing it to be configured to do so, for the restore options since this is ultimately being implemented for an endpoint you won't be able to restore from.
> -- @NickLaMuro 

Further clarifying this comment:

For our current use case, this option should be displayed for `backup` and `dump`, but should be hidden for the `restore` menu.
There are 2 possible solutions so far for configuration:

### the dry solution
@NickLaMuro suggests dry. It is no fun to have duplicate lists - duplication tends to go wrong quickly.

We display `menu_order` for all entries, unless an entry has `enable_for` key and the key does not include the current menu name (i.e.: dump, backup, restore).

```yaml
---
  database_admin:
    menu_order:
    - local
    - ftp://ftp.example.com/incoming/
    local: Local file
    prompts:
      ftp.example.com:
        enable_for: [dump, backup]
        filename_text: "The case number dash (-) filename. (e.g.: 12345-db.backup)"
        filename_validator: "^[0-9]{4,}-.+"
```

### the less logic solution
@kbrock suggests duplicating but removing special parameters that negate the list. Duplication stinks, but conditional logic is worse.

If `{restore, backup, dump}_menu_order` is available, display that. Otherwise fall back to display `menu_order`.

```yaml
---
  database_admin:
    menu_order:
    - local
    - ftp://ftp.example.com/incoming/
    restore_menu_order:
    - local
    - ftp://ftp.example.com/incoming/
    local: Local file
    prompts:
      ftp.example.com:
        filename_text: "The case number dash (-) filename. (e.g.: 12345-db.backup)"
        filename_validator: "^[0-9]{4,}-.+"
```

kbrock commented:
added rake_options and made custom endpoints enabled for various options (backup, dump, restore)

Does this work for you now @NickLaMuro ?

NickLaMuro commented on lib/manageiq/appliance_console/database_admin.rb:
```diff
@@ -140,21 +138,22 @@ def ask_ftp_file_options
-        hostname     = URI(server_uri).host
-        uri_filename = ask_custom_prompt(hostname, 'filename', "Target filename (e.g.: #{sample_case})")
-        @uri         = server_uri.gsub(sample_case, uri_filename)
+        hostname  = URI(server_uri).host
```
This is minor:

I found this a bit confusing to me for a bit since I saw you parsing out the `host` from the `server_uri`, and thought that was being being assigned to `@uri`.  Turns out that wasn't the case, but we also aren't using the `hostname` in any meaningful fashion except as a key.  Seems like we could just as easily do this:

```yaml
  database_admin:
    menu_order:
    - local
    - ftp://ftp.example.com/incoming/
    local: Local file
    prompts:
      ftp://ftp.example.com/incoming/:
        enabled_for: ["backup", "dump"]
        ...
```

And achieve the same affect by just passing around `server_uri` as the key.

* * * 

Again, this is minor, just thought it was a bit confusing from both a code and configuration perspective since there is two different "keys" being passed around.

kbrock replied:
Lets keep this as the host for now.
Using the full path would simplify the code and may be more clear
But I like reading the short prompt version

NickLaMuro replied:
Works for me.

If the above wasn't clear to on lookers, my above critique is pretty subjective, and I don't hold a hard opinion on this.

NickLaMuro commented on spec/database_admin_spec.rb:
```diff
@@ -752,48 +954,103 @@ def confirm_and_execute
+        it "displays custom ftp option with enabled array" do
+          expect(I18n).to receive(:t).with("database_admin.menu_order").and_return(%w(local ftp://example.com/inbox/filename.txt))
+          expect(I18n).to receive(:t).with("database_admin.local").and_return("The Local file")
+          expect(I18n).to receive(:t).with("database_admin.prompts", :default => {}).at_least(:once).and_return(:"example.com" => { :enabled_for => %w(restore backup) })
```
_**Pedantic**_:  Wouldn't this be a known quantity of `.once` or `.twice`?  Would save some characters and would be more explicit with those since this shouldn't be a variable number of calls as far as I can tell.

NickLaMuro replied:
Also, since this line is pretty long, could the prompt options be saved in a `let(:prompt_options)` to allow this to be more visable?

kbrock replied:
The actual value for `enabled_for` are different for most tests. I'll make a helper for this.

I'll make the `twice` more explicit

NickLaMuro commented on spec/database_admin_spec.rb:
```diff
@@ -752,48 +954,103 @@ def confirm_and_execute
+
+        it "displays custom ftp option with enabled array" do
+          expect(I18n).to receive(:t).with("database_admin.menu_order").and_return(%w(local ftp://example.com/inbox/filename.txt))
+          expect(I18n).to receive(:t).with("database_admin.local").and_return("The Local file")
```
_**Pedantic**_:  These above two seem like they could be tossed in a `before` hook to save them being defined twice.

NickLaMuro commented on spec/database_admin_spec.rb:
```diff
@@ -752,48 +954,103 @@ def confirm_and_execute
+          PROMPT
+        end
+
+        it "hids custom ftp option with disabled string" do
```
Typo:  `s/hids/hides`

NickLaMuro commented on spec/database_admin_spec.rb:
```diff
@@ -67,48 +67,175 @@
       end
+
+      # this is the complete implementation. the other 2 are paired down version of this
+      context "with localized file upload" do
```
_**Pedantic**_:  maybe "with I18n custom menu configs"?

If changed, make sure to change in the other two spots.

kbrock commented:
@miq-bot tag hammer/yes

carbonin commented:
@kbrock The console is actually a released gem so at some point when the backup/dump work stabilizes, we'll cut a release and include that in the hammer branch of manageiq-appliance
